Feature/openssl ecc keypair support

dol opened the pull request:
The current ECC (eliptic curver crypthography) support of the OpenSSL extensions has no possibility to generate a EC public key pair.

The PR is a modified version of the patch provided in https://bugs.php.net/bug.php?id=61204 .
- Renamed 'ec_group_name' parameter to 'curve_name' to be compliant with the export parameter naming of [openssl_pkey_get_details()](https://github.com/php/php-src/commit/5d1cc7d13536d1753adce40b903d71c5c841c30a#diff-69bad938d17f4283faa5f7fea17fa627R3512).
- Added some missing memory free statements
- Added more test cases

The PR also fixes some tabs/spaces formatting issues.

bukka commented on ext/openssl/openssl.c:
```diff
@@ -975,6 +990,18 @@ static int php_openssl_parse_config(struct php_x509_request * req, zval * option
+	/* set the ec group curve name */
+	req->curve_name = 0;
+	if (optional_args && (item = zend_hash_str_find(Z_ARRVAL_P(optional_args), "curve_name", sizeof("curve_name")-1)) != NULL
+		&& Z_TYPE_P(item) == IS_STRING) {
```
could you possibly add extra indent tab when continuing condition - it's a bit more readable then... ;)

dol replied:
I searched for all the '/^\s+&&/' occurrences within openssl.c and the indentation is always one tab. I leave it like this to be compliant with the file. @bukka Can you refer to a style guide that defines this case. I was not able to find one. Thank you.

bukka replied:
yeah leave it. it was just a suggestion to make it a bit more readable - not really defined anywhere. you are right that it's better to make it consistent with the rest of the file...

bukka commented on ext/openssl/openssl.c:
```diff
@@ -975,6 +990,18 @@ static int php_openssl_parse_config(struct php_x509_request * req, zval * option
 	PHP_SSL_CONFIG_SYNTAX_CHECK(extensions_section);
+#ifdef HAVE_EVP_PKEY_EC
+	/* set the ec group curve name */
+	req->curve_name = 0;
```
NID_undef would be better here as it's 0...

dol replied:
Thx for the advice. Makes sense.

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3644,6 +3697,68 @@ PHP_FUNCTION(openssl_pkey_new)
+						&& Z_TYPE_P(x) == IS_STRING
+						&& (y = zend_hash_str_find(Z_ARRVAL_P(data), "y", sizeof("y") - 1)) != NULL
+						&& Z_TYPE_P(y) == IS_STRING) {
+						EC_POINT *pnt = EC_POINT_new(group);
```
We should check if pnt gets allocated successfuly : `if (pnt) ...`

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3644,6 +3697,68 @@ PHP_FUNCTION(openssl_pkey_new)
+						&& (y = zend_hash_str_find(Z_ARRVAL_P(data), "y", sizeof("y") - 1)) != NULL
+						&& Z_TYPE_P(y) == IS_STRING) {
+						EC_POINT *pnt = EC_POINT_new(group);
+						EC_POINT_set_affine_coordinates_GFp(group, pnt, BN_bin2bn((unsigned char*) Z_STRVAL_P(x), Z_STRLEN_P(x), NULL),
```
Think that the ret code should be check here as well. It looks to me that this can fail in same case (would have to look properly but think we should check ret code anyway)

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3644,6 +3697,68 @@ PHP_FUNCTION(openssl_pkey_new)
+						EC_POINT *pnt = EC_POINT_new(group);
+						EC_POINT_set_affine_coordinates_GFp(group, pnt, BN_bin2bn((unsigned char*) Z_STRVAL_P(x), Z_STRLEN_P(x), NULL),
+							BN_bin2bn((unsigned char*) Z_STRVAL_P(y), Z_STRLEN_P(y), NULL), NULL);
+						EC_KEY_set_public_key(eckey, pnt);
```
Indent and check return code - it's using `EC_POINT_dup` so it can theoretically fail...

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3644,6 +3697,68 @@ PHP_FUNCTION(openssl_pkey_new)
+					}
+
+					if (!EC_KEY_check_key(eckey)) {
+						EC_KEY_generate_key(eckey);
```
Again check return code. I see that you do extra check_key later but think that if generate success, then you don't need to do that maybe (would need to check...). If so then this would be better:
```c
if ((EC_KEY_check_key(eckey) || EC_KEY_generate_key(eckey)) && EVP_PKEY_assign_EC_KEY(pkey, eckey)) {
   RETURN_RES(zend_register_resource(pkey, le_key));
}
```
instead of the bellow cond...

bukka commented on ext/openssl/tests/ecc.phpt:
```diff
@@ -0,0 +1,112 @@
+<?php
+
+$args = array(
+  "curve_name" => "secp384r1",
```
Can we use tabs for indent in tests? I know that it's a bit messy and the tests have mixed indent but it would be great to use just tabs. I will try to fix other tests indent later in the same way as I did in json ext... So please just tabs in all tests...

bukka commented:
First of all thanks for this contribution!

I cherry-picked the first commit with white space issues (48ae925) so pls merge master into it and it should hopefully reduce the diff (not sure why it's not picked up automatically... :) )

I had a quick look and added few comments. I didn't have time to verify and test it. I think that this would be a nice addition to 7.1 so we still have some time to do that later... ;) But looks good overall!

Thanks

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5142,6 +5275,34 @@ PHP_FUNCTION(openssl_get_cipher_methods)
 }
 /* }}} */
 
+/* {{{ proto array openssl_get_curve_names([bool aliases = false])
```
no aliases should be probably here... :)

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5142,6 +5275,34 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+	}
+
+	for(i = 0; i < len; i++)
+	{
```
CS - add space after for and put bracket on the same line...

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5142,6 +5275,34 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+	for(i = 0; i < len; i++)
+	{
+		sname = OBJ_nid2sn(curves[i].nid);
+		if (sname == NULL)
```
missing braces (this is required in our coding style even if it's not needed here by C standard...)

bukka replied:
I would probably do this instead:
```c
if (sname != NULL) {
    add_next_index_string(return_value, sname);
}
```

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5142,6 +5275,34 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+	size_t len = EC_get_builtin_curves(NULL, 0);
+	curves = emalloc((int)(sizeof(EC_builtin_curve) * len));
+	if (!EC_get_builtin_curves(curves, len)) {
+		return;
```
you should probably use `RETURN_FALSE` instead

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5142,6 +5275,34 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+
+	array_init(return_value);
+	size_t len = EC_get_builtin_curves(NULL, 0);
+	curves = emalloc((int)(sizeof(EC_builtin_curve) * len));
```
why do you cast it to int - `_emalloc` accepts `size_t`.

dol commented on ext/openssl/openssl.c:
```diff
@@ -3644,6 +3697,68 @@ PHP_FUNCTION(openssl_pkey_new)
+					}
+
+					if (group == NULL) {
+						php_error_docref(NULL, E_WARNING, "Unknown curve_name");
```
It makes no sense triggering an warning but no exiting at this stage. If there is no group parameters we don't have any change to create a EC key pair. I'll have to change this behavior.

dol commented:
@bukka Thank you for your time to review the PR and add very helpful comments and thoughts. As I mentioned in my code comment I found a inconsistency with the group parameters. In retrospective taking the patch from https://bugs.php.net/bug.php?id=61204 and fixing a few things was not enough. ;-)

I'll update the PR that fixes this additional found issues:
* Implement the changes @bukka suggested
* Return false for openssl_pkey_new() the curve name and the corresponding group parameters are missing or unknown
* If one of ['ec']['x'], ['ec']['y'] or ['ec']['d'] parameters missing when using openssl_pkey_new() softfail and generate a new EC key pair. That's already the case, but there is not test case for this. I'll add them to the ECC test case.
* Perform proper checks in openssl_pkey_new() and do cleanup if something fails.

I appreciate the cherry-pick of the white space cleanup.

dol commented:
Fixed the mentioned issues, rebased against current php-src master and squashed commits.

The PR is not complete yet. I found a incorrect handling of the private and public key parameters when creating a key from 'd', 'x' and 'y'.
I'll release a fix and the corresponding test cases soon.

dol commented:
From my side the RP is finished. This version of the PR fixes the following problem:
If a EC key was created by specifying the 'd', 'x', but 'y' was missing, a new key pair was generated due to the fact that the public key was missing. Event if 'x' and 'y' are missing, the point of the public key can be constructed (calculated) from the private key information 'd'.
This issue is fixed and covered by multiple test cases.

bukka commented:
Thanks. I will take a look a look once I have time and will test it. As I said there is quite a bit of time before 7.1 and need to sort out few other things before that.

Thanks again for your work!

dol commented:
Good to know. I'll provide an other patch to openssl_csr_new() that will help to define the content that goes into the x509 extension 'subject alternative name'.

bwoebi commented:
In general this looks like a self-contained addition, potentially eligible for 7.0.next too. So @bukka you may as well review and merge it now.

bukka commented:
I'm more in favour of 7.1 for this. It adds a new function and new options for pkey. It is also touching the current code in some places. We should also review it carefully to make sure that everything is fine. Anyway I will take a look once I have more time and finish all other stuff that I plan to do.

dol commented:
@bukka I just realized PHP 7.1 alpha1 was release some days ago. I updated the PR with the latest master.
Do I need a RFC to be passed to merge this enhancement? I've only contributed minor patches and fixes to the core.
Is this even a feature? Reading the https://wiki.php.net/rfc and additional source not helped to get a clear view. I found @bukka RFC about [OpenSSL AEAD support](https://wiki.php.net/rfc/openssl_aead) which is kind of a similar extension.
Thank you in advance for help.

bukka commented:
That's cool. I actually planned to look at this next weekend. Will have to merge few other things before that.

I think that we don't need an RFC if no one objects against the future. The AEAD had RFC because we didn't fully agree on API but this is much more straight forward. However it should be definitely introduced on internals. So if you could send an email summarising what exactly is added (new function, extra params for pkey_new...), that would be great. If no one objects for a week or so and all is cool, I will try to merge it before beta.

bukka commented on ext/openssl/tests/029.phpt:
```diff
@@ -0,0 +1,254 @@
+$detailsFromScratch9 = array(
+	"ec" => array(
+		"curve_name" => "prime192v1",
+		"d" => gmp_export("3138550867681922400546388175470823984762234518836963313664"),
```
Just a quick thing that I noticed in your email. Please don't use gmp in openssl tests as it creates dependency on gmp ext which would require extra skip if. Please use hex value in hex2bin.

dol replied:
Will do.

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3714,6 +3767,97 @@ PHP_FUNCTION(openssl_pkey_new)
+#ifdef HAVE_EVP_PKEY_EC
+		} else if ((data = zend_hash_str_find(Z_ARRVAL_P(args), "ec", sizeof("ec") - 1)) != NULL &&
+			Z_TYPE_P(data) == IS_ARRAY) {
+			pkey = EVP_PKEY_new();
```
C89 compat issue (should be done after definitions)

bukka commented on ext/openssl/openssl.c:
```diff
@@ -4059,11 +4207,25 @@ PHP_FUNCTION(openssl_pkey_get_details)
 					ASN1_OBJECT_free(obj);
 				}
 
+				const EC_POINT *pub = EC_KEY_get0_public_key(ec_key);
```
C89 - definition needs to be at the beginning of the block

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5213,6 +5375,33 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+	const char *sname;
+	int i;
+
+	array_init(return_value);
```
it should be done after EC_get_builtin_curves otherwise there will be memleak in case it failed (RETURN_FALSE...)

bukka commented on ext/openssl/openssl.c:
```diff
@@ -5213,6 +5375,33 @@ PHP_FUNCTION(openssl_get_cipher_methods)
+	size_t len = EC_get_builtin_curves(NULL, 0);
+	curves = emalloc(sizeof(EC_builtin_curve) * len);
+	if (!EC_get_builtin_curves(curves, len)) {
+		RETURN_FALSE;
```
curves should be freed

bukka commented on ext/openssl/openssl.c:
```diff
@@ -975,6 +990,18 @@ static int php_openssl_parse_config(struct php_x509_request * req, zval * option
+		&& Z_TYPE_P(item) == IS_STRING) {
+		req->curve_name = OBJ_sn2nid(Z_STRVAL_P(item));
+		if (req->curve_name == NID_undef) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unknown elliptic curve (short) name %s", Z_STRVAL_P(item));
```
remove TSRMLS_CC from here and everywhere else ;)

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3499,6 +3526,26 @@ static EVP_PKEY * php_openssl_generate_private_key(struct php_x509_request * req
+			case OPENSSL_KEYTYPE_EC:
+				{
+					if (req->curve_name == NID_undef) {
+						php_error_docref(NULL TSRMLS_CC, E_WARNING, "Missing configuration value: 'curve_name' not set");
```
TSRMLS_CC as above...

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3714,6 +3767,97 @@ PHP_FUNCTION(openssl_pkey_new)
+
+					if (pnt != NULL) {
+						if (!EC_KEY_set_public_key(eckey, pnt)) {
+							goto clean_exit;
```
In case we get here, it will result to double free of the group (ptr is not null but it's been already freed). I'd suggest to put `EC_GROUP_free(group);` after the condition.

bukka commented on ext/openssl/openssl.c:
```diff
@@ -3714,6 +3767,97 @@ PHP_FUNCTION(openssl_pkey_new)
+						}
+						EC_POINT_free(pnt);
+					}
+
```
I mean put `EC_GROUP_free(group);` here

bukka commented:
I just merged https://github.com/php/php-src/pull/1950 . If you find a little bit of time, it would be great if you could add `php_openssl_store_errors();` after each case that stores errors to the OpenSSL queue (of course relating just to your additions...)

bukka commented:
@dil ^

php-pulls commented:
**Comment on behalf of bukka at php.net:**

Superseded by #1959

dol commented:
@bukka You are a hero. I just wanted to implement your suggestions today. Haven't had the time to look into the improvement due to family time.
